datasource/opennms-direct/src/main/java/org/opennms/oce/datasource/opennms/jvm/DirectInventoryDatasource.java

```diff
@@ -28,58 +28,172 @@
+        this.alarmDao = Objects.requireNonNull(alarmDao);
     }
 
+    /**
+     * On init we will populate the inventory by retrieving all current inventory from the {@link NodeDao} and
+     * {@link AlarmDao}.
+     */
     public void init() {
-        /// Populate the map with the current set of inventory
+        nodeDao.getNodes().forEach(n -> inventoryObjects.addAll(Mappers.toInventory(n)));
+        alarmDao.getAlarms().forEach(a -> processAlarm(a, false));
+        initLock.countDown();
+    }
+
+    /**
+     * Wait for {@link #init the init} to finish to delay callbacks from being processed and clients from polling for
+     * inventory.
+     */
+    private void waitForInit() {
+        try {
+            initLock.await();
+        } catch (InterruptedException ignore) {
+            LOG.debug("Interrupted while waiting for init lock.");
+            Thread.currentThread().interrupt();
+        }
+    }
+
+    /**
+     * Adds the inventory resulting from handling a callback.
+     *
+     * @param inventoryObjects the inventory mapped from the callback
+     * @param notifyHandlers   whether or not to notify handlers
+     */
+    private void addInventory(List<InventoryObject> inventoryObjects, boolean notifyHandlers) {
         rwLock.writeLock().lock();
         try {
-            inventoryObjects.clear();
-            nodeDao.getNodes().forEach(n -> inventoryObjects.addAll(Mappers.toInventory(n)));
+            // Only add and notify for inventory that was not already known
+            Set<InventoryObject> newInventory = Sets.difference(this.inventoryObjects, new HashSet<>(inventoryObjects));
+
+            if (this.inventoryObjects.addAll(newInventory)) {
+                if (notifyHandlers) {
+                    inventoryHandlers.forEach(handler -> handler.onInventoryAdded(newInventory));
+                }
+            }
         } finally {
             rwLock.writeLock().unlock();
         }
     }
 
+    /**
+     * Process an alarm depending on if it is associated with a node or not.
+     *
+     * @param alarm       the alarm to process
+     * @param waitForInit whether or not to wait for init to finish first
+     */
+    private void processAlarm(Alarm alarm, boolean waitForInit) {
+        if (waitForInit) {
+            waitForInit();
+        }
+
+        Node nodeForAlarm = alarm.getNode();
+
+        if (nodeForAlarm != null) {
+            addInventory(Mappers.toInventory(nodeForAlarm), waitForInit);
+        }
+
+        addInventory(Mappers.toInventory(alarm), waitForInit);
+    }
+
+    @Override
+    public void handleAlarmSnapshot(List<Alarm> alarms) {
+        alarms.forEach(alarm -> processAlarm(alarm, true));
+    }
+
+    @Override
+    public void handleNewOrUpdatedAlarm(Alarm alarm) {
+        processAlarm(alarm, true);
+    }
+
+    @Override
+    public void handleDeletedAlarm(int alarmId, String reductionKey) {
+        // TODO: Anything needed here?
```

Comment:
We should delete any inventory that was derived from the alarm. Note that we should take care to avoid problems like the ones we ran into in the Kafka datasource (the reason for which we added reference counting).